File: nav_goal.cpp
#include <ros/ros.h>
#include <move_base_msgs/MoveBaseAction.h>
#include <actionlib/client/simple_action_client.h>
#include <iostream>
#include <geometry_msgs/Twist.h>
using namespace std;

typedef actionlib::SimpleActionClient<move_base_msgs::MoveBaseAction> MoveBaseClient;

void poseCallback(const geometry_msgs::Twist& amcl_pose)
{
  ROS_INFO("Thor is at ([%f], [%f])", amcl_pose.linear.x, amcl_pose.linear.y);
  std::cout << "Twist Received. " << std::endl;
}

int main(int argc, char **argv){

  ros::init(argc, argv, "map_nav_goal");

  //tell the action client that we want to spin a thread by default
  MoveBaseClient ac("move_base", true);

  //wait for the action server to come up
  while(!ac.waitForServer(ros::Duration(5.0))){
    ROS_INFO("Waiting for the move_base action server to come up");
  }

  move_base_msgs::MoveBaseGoal goal;
  //we'll send a goal to the robot to move 0.5 meter forward
  goal.target_pose.header.frame_id = "map";
  goal.target_pose.header.stamp = ros::Time::now();
  cout << "Frame: " << goal.target_pose.header.frame_id << "\n";
  float x , y; 
  cout << "Enter x coordinate: ";
  cin >> x; 
  cout << "Enter y coordinate: ";
  cin >> y;  
  goal.target_pose.pose.position.x = x;
  goal.target_pose.pose.position.y = y;
  goal.target_pose.pose.orientation.w = 1.0;

  ROS_INFO("Sending goal");
  ac.sendGoal(goal);

  ac.waitForResult();

  if(ac.getState() == actionlib::SimpleClientGoalState::SUCCEEDED)
    ROS_INFO("Moved to destination");
  else
    ROS_INFO("Failed to move to position");
  ros::NodeHandle n;
  ros::Subscriber sub = n.subscribe("/amcl_pose", 1000, poseCallback); 

  return 0;
  ros::spin();
}

File: nav_goal.py
#!/usr/bin/python

import roslib 
import rospy
#import simple action client
import actionlib
#imports messages used by move_base_msgs 
import move_base_msgs.msg
from geometry_msgs.msg import PoseWithCovarianceStamped

def amclpose(msg):
	print str(msg.pose.pose)
	global thor_state
	thor_state = [msg.pose.pose.position.x, msg.pose.pose.position.y, msg.pose.pose.orientation.w] 

def nav_goal_client(goal, client, goal_x, goal_y, goal_w):
	goal.target_pose.header.stamp = rospy.get_rostime()
	goal.target_pose.pose.position.x = goal_x
	goal.target_pose.pose.position.y = goal_y 
	goal.target_pose.pose.orientation.w = goal_w
	#sends goal to action server 
	client.send_goal(goal)
	#waits for the server to finish performing action 
	client.wait_for_result()
	return client.get_result()

def dist_offset(x_des, y_des, x_actual, y_actual):
	x_offset=x_des-x_actual
	y_offset=y_des-y_actual
	print "x offset: ", str(x_offset), "y offset: ", str(y_offset)
	return (x_offset, y_offset)

def accurate_nav(threshold):
	client=actionlib.SimpleActionClient("move_base",move_base_msgs.msg.MoveBaseAction)
        client.wait_for_server()
        #creates goal to send to action server
        goal=move_base_msgs.msg.MoveBaseGoal()
        goal.target_pose.header.frame_id = "map"
        print "current frame: ", goal.target_pose.header.frame_id
        global x, y, w
        x = float(input("x coordinate: "))
        y = float(input("y coordinate: "))
        w = float(input("orientation: "))
	(x_offset, y_offset) = (1, 1)
	while abs(x_offset) > threshold and abs(y_offset) > threshold:
		result = nav_goal_client(goal, client, x, y, w)
		print "updated position" 
		rospy.Subscriber('amcl_pose',PoseWithCovarianceStamped,amclpose)
		rospy.sleep(1)
		(x_offset, y_offset) = dist_offset(x, y, thor_state[0], thor_state[1])
		x = x + x_offset
		y = y + y_offset
	return True

if __name__ == '__main__':
	try: 
		rospy.init_node('nav_goal')
		task = accurate_nav(0.1)
		print "arrived!"	

	except rospy.ROSInterruptException:
		print "program interrupted before completion"

